### ISSUE TITLE
Improve settings editor accessibility

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#97567]
Issues about the accessibility of the settings editor UI have come up a few times, and I'd like to collect more info about how and whether screenreader users are using it, what issues they are having, and how it could be made to work better.

- Is the document structure of the settings editor easy to understand and navigate around?
- Can users navigate to different setting categories, search for particular settings, or browse through the list efficiently?
- Can setting names, descriptions, and values be read consistently?
- Does editing different types work well?
- Does the JSON editor work well for settings that have to be edited that way?
- Does the JSON editor in general work better for screenreader users than the UI?
  - If so, should it be the default? If not, would it be possible to make that the case?

A few issues with discussions about settings editor accessibility:

- https://github.com/microsoft/vscode/issues/92485
  - Mentions that tab navigation in the virtualized list doesn't work well. We suggest using JSON editor
- https://github.com/microsoft/vscode/issues/92488
  - Bug with orca not speaking categories
- https://github.com/microsoft/vscode/issues/71952
  - Wrong screenreader mode while editing
- https://github.com/microsoft/vscode/issues/66237
  - Keyboard navigation issues (page down, spacebar aren't supported) and discussion about row selection
- https://github.com/microsoft/vscode/issues/97068
  - The setting scope tabs do not have proper aria attributes
- https://github.com/microsoft/vscode/issues/94525
  - Difficulty understanding the relationship between the "table of contents" tree and the position in the settings list. Pointing out VS, Chrome, and FF as better examples.

**Comment by roblourens** [comment]
Got some great feedback in today's call, thanks for everyone who joined. I will flesh out my notes after I listen to the recording but here are some quick notes

- People would prefer to use the settings GUI editor - the JSON editor is necessary but we can't rely on it to be the sole "accessible option"
- People get lost in the page structure, it's not clear how the category list relates to the settings list on the right
- You can tab from the category list to the setting list, but you can't tab back
  - You are stuck shift+tabbing to the top of a long list, and also focus is trapped on the focus sink at the top of the list (bug, there is an issue for this)
- There is no indication when you have tabbed across a category boundary, you are just tabbing through settings forever, and this is very confusing and unexpected
- Proposal that seemed generally popular: instead of showing an extremely long virtualized settings list on the right, show a shorter non-virtualized list which is only scoped to the selected category on the left. This would mean that tabbing could work more normally, and other navigation modes like "browse mode" in NVDA should work as expected. This would make the settings editor work more like a typical **form**.
- Another proposal: make settings rows selectable, making the settings list work more like a typical **list**. Navigate using arrow keys, maybe left/right to go between the category list and the settings list.

**Comment by alex19EP** [comment]
ok. lets summarize. 📒

## main points

* settings shouldn't be an endless list.
    * need to limit them to one category.
* It should be impossible to switch to another category from the settings, only from the list of categories.

## settings navigation 

at the call, opinions were divided. I will speak for myself.

* Navigation in the settings list should be done with arrows. the tab should go to the setting or button wich relates to current setting.

**Comment by alex19EP** [comment]
oh @roblourens  was faster than me. ☺

**Comment by webczat** [comment]
Generally the arrow navigation proposal is not typical as in, for me the typical workflow is using tabs. On the other hand I don't think the idea with exposing settings as an arrowable list is necessarily a bad idea, but different and not sure if obvious to someone new. At least it shouldn't be confusing.

**Comment by alex19EP** [comment]
@webczat i think arrows will be cool because in code settings we have not only settings but for example changed indicator or button with copy json or clear actions.

**Comment by webczat** [comment]
actually makes sense. currently we can't tab to more actions, we have to shift+f9 or use browse mode. maybe that's really better.

**Comment by roblourens** [comment]
I think the arrow key model would be confusing and non-standard - we actually used this model in an early version of the settings editor, and it got bad reviews. I think it's confusing to pick a setting with the arrow keys, then tab _into_ that setting, but then in that context, the arrow keys have different behaviors, for example they will move the cursor around in an input box or expand a dropdown. Then to pick the next setting, you have to tab out of there, back to the setting level, then use arrow keys, and so on. This is just not a typical interaction model and I don't think I've seen something like that before. 

To me, the settings editor is fundamentally a form and it will be easiest for new users to understand if it can be navigated like other forms.

**Comment by webczat** [comment]
maybe, although I could live with this. vscode does have some nonstandard behaviors here and there and I know I would not have problem with this one, so I'm open to both ways. :)

**Comment by alex19EP** [comment]
> I think it's confusing to pick a setting with the arrow keys, then tab into that setting, 

we can use enter for that. tab will be move to settings button menu.

I think the navigation using the arrows will be faster.
tab and shift tab will move you to category and settings trees, setting button menu. and such.
in this way we show the overall composition of this window.

also do not forget that the list of settings can be quite long.

**Comment by isidorn** [comment]
@roblourens is it an easy change to change the settings tree to use the arrow key model? If yes, maybe you can just produce a build so we can try it out and provide feedback. If you got reference to those bad reviews from before that might also be helpful as I might be missing something which our users hated with this arrow model.

**Comment by roblourens** [comment]
I think it would take some effort to get it right. I am mainly just thinking of internal feedback in the UX call and offline. There was this issue: https://github.com/microsoft/vscode/issues/54039 but it doesn't capture everything.

@alex19EP I didn't fully get your point here:

> i think arrows will be cool because in code settings we have not only settings but for example changed indicator or button with copy json or clear actions.

How do those other things relate to the arrows?

**Comment by alex19EP** [comment]
> How do those other things relate to the arrows?
I think that it would be more logical to move to this button using the tab. I hope it’s clear now.

**Comment by roblourens** [comment]
I understand why this scheme would have some benefits, but I think that what really bugs me is the idea that a new user would see this page which looks like a normal form with input boxes and dropdowns and things, and not be able to tab to those things, like they can pretty much 100% of the time in other places. Instead they try to tab, fail, then have to discover this new way of interacting with the page.

**Comment by isidorn** [comment]
@roblourens thanks for thinking about this. Here's a practical idea, just as food for thought:
Keep current tab behavior in general so existing users are not broken.
Just change Arrow behavior. Here's how the arrow behavior would be:
* Up / down to navigate through elements. 
* Space / enter to "move inside the element" 
* Once you are inside the element up/down are specific to the element, thus they do not navigate top level
* once you are inside the element: `esc` to move outside of the element
* While focus outside of the element typing characters searches and filters

**Comment by roblourens** [comment]
Maybe we can experiment with it but I guess it feels like we are reinventing something that a screenreader should help users do already, and if I understood screenreaders better, I would know how this "should" work without creating a new control scheme.

**Comment by jhomme** [comment]
> 
> 
> > I think it's confusing to pick a setting with the arrow keys, then tab into that setting,
> 
> we can use enter for that. tab will be move to settings button menu.
> 
> I think the navigation using the arrows will be faster.
> tab and shift tab will move you to category and settings trees, setting button menu. and such.
> in this way we show the overall composition of this window.
> 
> also do not forget that the list of settings can be quite long.


Is this something like File > Options in Microsoft Office? if so, it is workable.

**Comment by isidorn** [comment]
@roblourens do we plan to look into this in May? I think we should push it out to June?

**Comment by isidorn** [comment]
Unassignig myself for now. @roblourens let me know how I can best help here.

**Comment by 9at8** [comment]
I am trying to make navigation easier by replacing the endless virtual list with smaller non-virtual lists for each setting group, which is the first proposal in this [comment](https://github.com/microsoft/vscode/issues/97567#issuecomment-627414347). I have an experimental build of VSCode ready to test. I would love to get your feedback on it!

Note that search in the settings ui has not been implemented in these experimental builds yet.

Here are the builds:

Windows x64: https://az764295.vo.msecnd.net/insider/785761f045eaf0a266eb3bba39400839d0059819/VSCode-win32-x64-1.48.0-insider.zip

Linux x64: https://az764295.vo.msecnd.net/insider/785761f045eaf0a266eb3bba39400839d0059819/code-insider-1594765830.tar.gz

macOS: https://az764295.vo.msecnd.net/insider/785761f045eaf0a266eb3bba39400839d0059819/VSCode-darwin-insider.zip

**Comment by jhomme** [comment]
Hi,

How do I run this along with my VS Code Insider that I now have?

 

Thanks.

 

Jim

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by 9at8** [comment]
@jhomme Here are the steps:

**Note**: Make sure that all windows of code insiders are closed (if you're using a mac, make sure you quit it using Cmd + Q)

### Windows:

- Download the zip file from the link and extract it.
- Run the binary "Code - Insiders.exe" that you just extracted.

### Linux:

- Download the tarball from the link.

Run these commands from the terminal:

- `tar -xf ./code-insider-1594765830.tar.gz`
- `./VSCode-linux-x64/bin/code-insiders`

### macOS:

- Download the zip from the link, and unzip it.
- Run this in the terminal: `./Visual Studio Code - Insiders.app` or open it from finder

**Comment by jhomme** [comment]
Hi,

Will this make me have two versions of VS Code Insider? If not, what will happen when the next update of the VS Code Insider I have runs tomorrow?

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by 9at8** [comment]
Nope, this will not replace your local installation of vscode. If you open vscode using the start menu / spotlight / etc you would be using the regular version of vscode that you have installed. It also **won't** interrupt the update process.

**Comment by 9at8** [comment]
It also won't any additional versions of VSCode

**Comment by jhomme** [comment]
Hi,

I apologize, because I feel like I’m spamming this thread. I’m still learning my way around GitHub. How would I search for accessibility issues involving VS Code on Mac with VoiceOver, so that I can start looking at that and putting comments where they should go?

 

Thanks.

 

Jim

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by jhomme** [comment]
Hi,

This result is on PC. I only have two requests. I like how it works. It’s much better than before. Just two asks.

 

*	Do not allow tabbing to go into the other parts of the program when the settings window is open. When the user hits ESCAPE, let tabbing go back to normal.
*	Sometimes the links to get more information about a feature come before a field and sometimes after. Make this placement consistent, but I prefer after.

Thank you so much for working on this. It is way better.

 

Jim

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by webczat** [comment]
I strongly disagree with any form of focus trapping, as in tab should move further, the same as it would in anything else except fields accepting a tab.

**Comment by jhomme** [comment]
Hi,

This is not focus trapping. You can hit Escape to get out. My proposal would act like any dialog you normally see, where you can dismiss it to get back to normal navigation.

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by webczat** [comment]
however, this is not a dialog. this is an editor. so, if that is not a multiline text field, you are expected to be able to tab out of the area without any problem. At least it is my expectation when being in settings that it is an editor, so if I can tab around it, I can tab out of it, like from a view.

**Comment by jhomme** [comment]
Hi,

The W3C specific success criterion is 2.1.1 and specifically states Tab or Escape.

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by webczat** [comment]
so, let's agree to disagree then, and maybe other people have their own thoughts. :)

**Comment by jhomme** [comment]
Hi,

Proposal. Since some of the setting areas have many fields, SHIFT+F6 to jump back to the tree view rather than needing to tab or shift tab many times.

 

Thanks.

 

Jim

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by jhomme** [comment]
Hi Michal,

This is coming from an international standards point of view. We need to do that rather than injecting personal preference when possible. This is not a my opinion versus yours argument. This needs to cover all disabilities, including people with motor disabilities who may not be able to use a keyboard. The fewer key strokes for them the better. 

 

Even though this area is not a dialog in the pure Windows sense, it simulates dialog behavior in that it is a discrete area of the application that users treat like a dialog. Think about Microsoft Office programs such as Outlook or Word, which you go into an area to set options. They are not dialogs, and they force you to use Escape, OK, or Cancel to get back to the main part of the application. 

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by webczat** [comment]
the original argument of mine is not quite related to preferences but to consistency. I mean, the current behavior is consistent with other similar parts of vscode. No existing part of vscode known to me reacts that way, like other editors. That is why I have doupts about this one. Or I just don't quite understand the benefit. If you want to close settings you press ctrl+w like for all other editors. If something like shift+f6 was added you would be able to quickly focus the tree view. If you want to switch from settings to anything else you use same shortcuts as for the other editors like alt+1 or so.

**Comment by 9at8** [comment]
The problem we're trying to fix with this new layout is that users have to repeatedly tab/shift+tab to navigate to the table of contents. The change in the experimental build also enables using Page up, Page down, Up arrow, and Down arrow keys to navigate as well.

> How would I search for accessibility issues involving VS Code on Mac with VoiceOver, so that I can start looking at that and putting comments where they should go?

@jhomme [This link](https://github.com/microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Aaccessibility) should give you a list of open issues related to accessibility. [This link](https://github.com/microsoft/vscode/blob/master/CONTRIBUTING.md#reporting-issues) explains more about reporting issues.

**Comment by 9at8** [comment]
Hey everyone!

As a consequence of making the list non-virtual, you can also use page up/page down to navigate the list of settings. But there's a performance penalty to this. :cry: Things start to slow down once you start rendering 100+ settings at once. I'm experimenting with pagination, infinite scrolling with a "load more items" button, and infinite scrolling with automatic updates. I would love to hear the concerns, feedback, pros, or cons of using these techniques.

**Comment by jhomme** [comment]
Hi,

My opinion is mostly there won’t be that many settings. Can we check out a build, then give an opinion?

 

Thanks.

 

Jim

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by 9at8** [comment]
@jhomme for sure! I'll post a build for the pagination version and the load more settings version tomorrow. Stay tuned! 🙂

**Comment by alex19EP** [comment]
ok. my feedback about test build.

> * Sometimes the links to get more information about a feature come before a field and sometimes after. Make this placement consistent, but I prefer after.

agree with @jhomme on this.

> Proposal. Since some of the setting areas have many fields, SHIFT+F6 to jump back to the tree view rather than needing to tab or shift tab many times.

I agree that we need a way to quickly move focus to the settings categories, but I'm not sure if it should be the f6 key.

> The change in the experimental build also enables using Page up, Page down, Up arrow, and Down arrow keys to navigate as well.

i tried that but didn't succeed. maybe I didn't understand something. @9at8  can you explain in more detail?

**Comment by 9at8** [comment]
Sorry for the delay!

Here's the build for the pagination version. There should be "Next" and "Previous" buttons at the bottom of the page if there are more settings available. Currently, the page size is set to 20, but we can change it if you think that would be better. Search is working in this build, so let me know if you have any feedback for that as well. Search is also an excellent place to try out pagination. Try using words that would yield a lot of results as a search query; for example: "a", "the,

Windows x64: https://az764295.vo.msecnd.net/insider/656350171fed8b830b4b73e9cf48068f67e11697/VSCode-win32-x64-1.48.0-insider.zip

macOS x64: https://az764295.vo.msecnd.net/insider/656350171fed8b830b4b73e9cf48068f67e11697/VSCode-darwin-insider.zip

Linux x64: https://az764295.vo.msecnd.net/insider/656350171fed8b830b4b73e9cf48068f67e11697/code-insider-1595353470.tar.gz

**Comment by 9at8** [comment]
> Sometimes the links to get more information about a feature come before a field and sometimes after. Make this placement consistent, but I prefer after.

@jhomme @alex19EP I don't think I understand this correctly. Can you give me an example of where this happens?

> I agree that we need a way to quickly move focus to the settings categories, but I'm not sure if it should be the f6 key.

We do have a keybinding that you can configure for that, but it doesn't have a default. If you open up "Preferences: Open Keyboard Shortcuts" from the command palette, and search for `settings.action.focusTOC` you should be able to configure it. Although, I would like to get recommendations for a default keybinding as well.

> i tried that but didn't succeed. maybe I didn't understand something. @9at8 can you explain in more detail?

I was wrong about the Up and Down arrow keys, but if you are using browse mode, you should be able to use Page up and Page down to navigate the page now.

**Comment by 9at8** [comment]
Here are the builds for the settings editor with a load more button in the bottom, instead of pagination controls. At the end of the settings page, you can find the "Load more settings" button. As before, Search is working in this build as well.

**Comment by jhomme** [comment]
Hi,
These observations may not answer the questions we are considering, but I wanted to list them here. These are all observed with NVDA, the newest release candidate build, which has improved responsiveness.

* Open Settings and navigate to **Commonly Used.**
* Tab once and you land on a link. Tab again, and you land on **Files Autosave**. Tab again and you land on another link. These two links, or seemingly 2 links, lead to the same place. I feel that the first one should not be there or that we should not be able to land on it.
* The other thing that happens here is that the longer description about getting more information about the setting is spoken when focus is on the combobox that allows the user to change the setting, rather than the link. You can't get information about the setting by pressing ENTER on the combobox. You have to hit it on a link.
* I noticed that if you change a setting, such as **Files AutoSave** and then press ENTER, you get reliable focus behavior. If you change a setting and hit TAB, focus lands lands at the top of the document and with the latest NVDA, Browse mode turns on. The Browse mode behavior is probably due to NVDA changing so that it does not automatically go into Application mode when VS Code runs. When you switch into Application mode with INSERT+SPACE, NVDA reports that focus is at the document level. If you then tab, you land back on the tree view.
* The behavior I expect is that if you tab, you should land on the next item in the order of the things to change, which in this case would be the second link, as I was talking about above.
* Another observed behavior is that any time I change a setting and press ENTER, I have to tab all the way through until I get back to where I left off. focus should probably return to the item I changed, so that I can quickly verify that I have made the correct change. I feel that it should be OK to have focus move to the next item when I change it and press TAB. It would be disconcerting to press TAB and have focus stay where it is.
* An example of the link about a setting coming before the setting is the one about **Editor Multicursor Modifier**. I again feel that you should encounter the setting, then the **Read more** link.
* When changing a setting, it would be a very nice thing to have if the editor would report that the setting you land on is the default. It reports that the setting is modified.
*

**Comment by jhomme** [comment]
Hi,

I think you do want focus to land on the links so you can get more information if you need to. 

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by jhomme** [comment]
Hi,
I would like to encourage you to release this in the Insider edition even though it is not perfect. It is much better than before.

**Comment by roblourens** [comment]
Thanks for the feedback! We've also explored the other option, navigating the settings list more like a list, using the arrow keys, with rows that can be focused. There is a more detailed description of it in this issue: https://github.com/microsoft/vscode/issues/104318 and links to builds. Would really appreciate any feedback. We will be trying this out in Insiders once we get the next release out.

**Comment by jhomme** [comment]
Hi,

Can’t wait to see this. 

 

Jim

 

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by roblourens** [comment]
Hey @alex19EP have you had a chance to try the new build in https://github.com/microsoft/vscode/issues/104318 yet? This essentially implements what we discussed way back in may

**Comment by isidorn** [comment]
I really love the improvments in VS Code insiders.
Though would also love to hear @jhomme and @alex19EP feedback

**Comment by isidorn** [comment]
@jhomme thanks for feedback, we are glad you think this is going in the right direction.
As for the aria-label and `aria-description`, mostly we just use `aria-label` since if I am not mistaken `aria-description` is not yet fully supported by all screen readers. Though it would be great if you create a new issue regarding this where we can continue this discussion and involve the appropriate screen reader folks.

**Comment by jhomme** [comment]
Hi,
I will test soon with NVDA and JAWS and try to create a new issue that
spells out the symptoms. I will use the JAWS ability to read out what
it thinks the tags and attributes are. It will take some time to write
it all up. No worries, though.

On 9/29/20, Isidor Nikolic <notifications@github.com> wrote:
> @jhomme thanks for feedback, we are glad you think this is going in the
> right direction.
> As for the aria-label and `aria-description`, mostly we just use
> `aria-label` since if I am not mistaken `aria-description` is not yet fully
> supported by all screen readers. Though it would be great if you create a
> new issue regarding this where we can continue this discussion and involve
> the appropriate screen reader folks.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/microsoft/vscode/issues/97567#issuecomment-700805906


-- 
==========

Jim Homme
Website: https://www.jimhomme.com/
Twitter: @jimhomme
Facebook: http://www.facebook.com/jimhomme
LinkedIn: https://www.linkedin.com/in/jimhomme
Skype: jim.homme
FreeChess: jhomme
